### ISSUE TITLE
FD leak on 3.0-dev11/12

### ISSUE THREAD

**Issue by felipewd** [haproxy/haproxy#2574]
### Detailed Description of the Problem

Hi,

We've been testing 3.0-dev11 and 12 in production and we've noticed on single node with huge amounts of FD usage (millions), getting to the point of widespread exhaustion.

We've collected some data that might help in `anon` mode. But the jist of it is (this after a restart):

```
07:43:57 PM   UID       PID threads   fd-nr  Command
07:43:57 PM   401    129505      64  266148  haproxy
```

according do netstat, most of the offending FDs seem to be on UNIX sockets, which is our backend:

```
# netstat -alnp  | grep TIME_WAIT | wc -l
5236
# netstat -alnp  | grep unix | wc -l
196715
```

We have nodes running dev8 that do not have this issue...but we can't be sure about dev9/10.

### Expected Behavior

Not leaking FDs :-)

### Steps to Reproduce the Behavior

Not really sure.

### Do you have any idea what may have caused this?

We've noticed that since dev8 there was some shutdown refactoring, so that might point to some possible explanations.

@rnsanchez flagged d3fc982cd78 after a quick look on our part.

We'll rollback to dev8 for the time being, but we're happy to help where we can to track this down.

### Do you have an idea how to solve the issue?

Not at all

### What is your configuration?

```haproxy
global 
    nbthread 64
    hard-stop-after 7d
    maxconn 900000
    ulimit-n 3150919
    user haproxy
    group haproxy
    daemon
    master-worker no-exit-on-failure
    expose-experimental-directives
    unix-bind user haproxy mode 777

    node haproxy-frontend

    localpeer peer_reload

peers peer_local
    peer peer_reload 127.0.0.1:10000

...
defaults
    mode    http
    retries 1
    maxconn 900000
    timeout connect 1s
    timeout server 100s
    timeout client 100s
    timeout http-request 6s
    timeout http-keep-alive 40s
    timeout tunnel 300s
    timeout client-fin 4s
    timeout server-fin 4s

    option splice-response


frontend SSL

    bind ipv4@:80 shards by-thread
    bind ipv6@:80 shards by-thread
...
<bunch of acl and stick-tables stuff that don't seem to be related>


    use_backend %[req.hdr(host),lower,map(/etc/haproxy/hyper.map)]


backend hyper
    http-reuse always
    server v01 abns@/tmp/hyper.sock
```

### Output of `haproxy -vv`

```plain
HAProxy version 3.0-dev12-d236b43 2024/05/18 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 6.1.62 #1 SMP PREEMPT_DYNAMIC Thu Nov 16 17:38:35 UTC 2023 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_NS=1 USE_PCRE=1 USE_PCRE_JIT=1
  DEBUG   = 

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT +PCRE_JIT +POLL +PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN -SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=128).
Built with OpenSSL version : OpenSSL 1.1.1w  11 Sep 2023
Running on OpenSSL version : OpenSSL 1.1.1w  11 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.6
Built with network namespace support.
Built with zlib version : 1.2.13
Running on zlib version : 1.2.13
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.45 2021-06-15
Running on PCRE version : 8.45 2021-06-15
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
We have `show fds` and `show sess` with anon mode.
```

### Additional Information

[fds.txt](https://github.com/haproxy/haproxy/files/15381251/fds.txt)
[sess.txt](https://github.com/haproxy/haproxy/files/15381252/sess.txt)

We also have some `sar` information but found nothing extraordinary these (except for the obvious FD usage) that might indicate where the problem might be.

**Comment by felipewd** [comment]
(just for reference we have 5M `nr_open` and we've reached that limit...though at this point we just killed haproxy and couldn't collect anything extra).

After downgrading to dev8:

```
07:58:14 PM   UID       PID threads   fd-nr  Command
07:59:53 PM   401    135377      64   18408  haproxy
```

which is on par with our backend program:

```
08:00:14 PM   UID       PID threads   fd-nr  Command
08:00:14 PM     0    191194    1619   20562  hyper
```

**Comment by einval22** [comment]
Hi Felipe ! 

Thanks a lot for reporting this !
Indeed, these patches were not tested under high load with UNIX sockets back-end.

I'm trying to reproduce on my side.

**Comment by einval22** [comment]
Hi Felipe !

Could you please, try this patch and give us a feedback:

```
--- a/src/proto_uxst.c
+++ b/src/proto_uxst.c
@@ -241,8 +241,10 @@ static int uxst_connect_server(struct connection *conn, int flags)
 
        /* perform common checks on obtained socket FD, return appropriate Stream Error Flag in case of failure */
        fd = conn->handle.fd = sock_create_server_socket(conn, be);
-       if (fd & SF_ERR_MASK)
+       if ((fd & SF_ERR_MASK) == fd) {
+               conn->handle.fd = -1;
                return fd;
+       }
 
        /* FD is ok, continue with protocol specific settings */
        if (global.tune.server_sndbuf)
```

Error in fd checks in proto_uxst.c  was introduced due to auto rebase :(, when I've refactored this part.
    
Generalized now sock_create_server_socket() function returns an appropriate Stream Error Flag in case, when we failed to create a socket in order to connect to backend. 

conn->handle.fd should be also set to -1 as well in this case, as it could be checked further.

Kind regards,

**Comment by felipewd** [comment]
hello @vkssv,

Thanks for the quick reply!

No luck with this patch. After restarting the daemon with only 4Gbps of traffic (less than 3 minutes uptime):

```
02:25:53 PM   UID       PID threads   fd-nr  Command
02:25:54 PM   401    254144      64   98115  haproxy

```

I'm trying to collect more useful information. So here it is `show activity`:

```
thread_id: 21 (1..64)
date_now: 1716301504.657942
uptime_now: 213.629005
ctxsw: 24467229 [ 438572 297876 352100 375259 476306 405122 351316 412928 362832 356865 352561 394156 436886 405923 443821 420953 462232 374077 348483 349365 296139 432299 431790 420263 492175 341812 253760 430391 262265 449007 447487 324935 367127 395589 465974 283269 281000 418299 447401 373076 391480 393007 342545 403674 376745 421906 487688 366666 414612 381408 330348 366014 509137 288780 325409 327713 249055 338594 394926 384386 405820 412817 313443 409365 ]
tasksw: 230153 [ 3621 3317 3484 3236 3706 3666 3314 3801 3933 3630 3562 3930 3500 3611 3807 3746 4026 3656 3462 3772 3661 3372 3511 3685 4177 3790 3429 3529 3662 3513 3763 3459 3393 3389 3944 3167 3611 3651 3774 3597 3293 3485 3116 3870 3463 3758 3645 3495 3784 3109 3870 3519 3443 3693 3532 3614 3387 3337 3880 3662 3620 3597 3391 3733 ]
empty_rq: 1196620 [ 19354 18103 18727 19431 19197 19083 18795 18981 18875 18060 18910 18797 19356 18425 19093 19654 18942 18969 18816 18597 18021 19743 18812 19171 19121 19188 17664 18856 18384 18531 19242 18538 18154 18687 19166 18183 18390 18753 18988 18713 18486 18400 18208 18301 18937 18531 19270 18656 18730 18752 18068 18630 19506 18196 17825 17910 18382 18751 18875 17689 18766 18167 18498 18616 ]
long_rq: 15857 [ 388 268 265 154 232 221 179 254 224 266 299 199 317 293 305 201 316 190 176 231 141 224 243 267 351 137 149 265 210 326 267 153 231 246 358 222 152 444 226 169 253 284 168 406 258 198 257 254 356 401 182 318 263 179 319 310 143 158 219 214 329 293 158 178 ]
curr_rq: 1 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
loops: 3335479 [ 55626 46009 47022 50979 59433 54681 47072 53079 48711 50090 51050 53879 56346 51930 56195 56944 57581 52871 48127 49982 43994 56260 54315 55200 61515 52017 42148 53522 41781 58797 56715 48635 53385 50698 62698 42933 42807 56800 57009 56975 51814 52282 48640 53150 50700 54920 64307 47902 56022 49478 46983 52329 61729 42309 47277 47507 41283 47767 53398 59232 52597 55503 47180 55359 ]
wake_tasks: 15860 [ 388 268 265 154 232 221 179 254 224 266 299 199 317 293 305 201 316 190 176 231 141 224 243 267 351 137 149 265 210 326 267 153 231 247 358 222 152 444 226 169 253 284 168 406 258 198 257 254 356 401 182 318 264 179 319 311 143 158 219 214 329 293 158 178 ]
wake_signal: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
poll_io: 2336410 [ 39922 30163 31214 35376 44069 39168 31460 37388 33023 34510 35298 38430 40769 36322 40611 41301 42023 37301 32598 34231 28304 40658 38686 39499 46046 36476 26474 37938 26091 43323 41059 33035 37826 35027 47183 27101 27073 41041 41489 41899 36186 36537 33128 37433 34783 39399 49210 32227 40436 33594 31195 36626 46133 26623 31362 31806 25535 32178 37865 44164 36856 40224 31522 39983 ]
poll_exp: 26 [ 0 0 0 6 0 5 0 0 0 0 0 0 0 0 2 0 5 0 2 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
poll_drop_fd: 33 [ 0 0 0 0 1 0 0 0 1 1 0 1 1 0 0 0 2 0 0 0 0 1 0 0 1 0 0 0 0 0 1 3 1 0 1 0 0 0 0 0 4 0 0 0 0 0 1 0 0 2 0 0 0 1 0 2 0 0 1 1 3 2 1 0 ]
poll_skip_fd: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
conn_dead: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
stream_calls: 105041 [ 1661 1547 1625 1460 1646 1675 1486 1848 1783 1643 1616 1817 1609 1669 1703 1695 1915 1668 1576 1751 1641 1474 1644 1722 1887 1730 1538 1579 1627 1609 1757 1564 1479 1552 1821 1412 1610 1701 1715 1573 1554 1605 1378 1807 1581 1702 1659 1655 1764 1410 1810 1588 1518 1629 1653 1701 1481 1504 1791 1716 1684 1693 1483 1647 ]
pool_fail: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
buf_wait: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
cpust_ms_tot: 1596 [ 35 19 23 33 46 41 19 33 33 19 19 34 27 19 29 19 19 21 0 19 35 19 19 19 19 19 19 19 19 36 19 19 31 32 53 36 19 31 23 26 19 27 23 2 32 19 19 19 22 27 21 19 34 51 30 19 19 29 19 29 31 19 29 19 ]
cpust_ms_1s: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
cpust_ms_15s: 6 [ 0 0 0 0 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0 0 ]
avg_cpu_pct: 1 [ 1 1 0 0 1 1 1 1 0 0 0 1 1 1 1 1 1 0 0 0 1 1 0 0 1 0 1 1 0 1 0 0 0 1 1 0 1 0 1 0 0 0 0 0 1 1 1 1 1 1 1 0 1 1 1 0 0 1 0 0 1 1 0 1 ]
avg_loop_us: 28 [ 31 31 25 26 28 27 31 27 27 35 25 30 30 32 34 32 28 27 30 24 36 26 30 26 33 25 29 25 26 26 25 27 32 32 26 29 25 32 24 20 28 26 20 21 31 29 30 34 30 25 30 29 22 29 29 25 27 32 27 28 32 28 22 27 ]
accepted: 50926 [ 823 715 740 736 801 793 764 800 873 795 849 812 847 811 826 789 852 811 810 775 792 772 781 804 827 800 795 749 847 791 832 764 745 786 798 839 805 802 779 807 754 809 739 829 786 834 762 756 824 754 829 784 836 815 788 810 770 781 779 799 762 777 789 823 ]
accq_pushed: 216 [ 4 4 4 4 4 4 4 4 4 4 4 4 5 4 4 4 4 4 4 4 4 3 3 3 3 3 3 3 3 3 3 3 3 3 3 4 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 4 3 3 ]
accq_full: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
accq_ring: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
fd_takeover: 259 [ 2 4 4 4 4 6 3 6 6 3 2 7 2 12 3 2 5 3 6 2 2 6 5 5 3 5 5 5 5 1 2 6 3 3 4 2 6 3 3 3 4 3 3 9 1 2 4 3 2 2 4 1 5 4 5 5 5 5 1 6 7 9 1 5 ]
check_adopted: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_started: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_active: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
check_running: 0 [ 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ]
```

show info
```
Name: HAProxy
Version: 3.0-dev12-d236b43
Release_date: 2024/05/18
Nbthread: 64
Nbproc: 1
Process_num: 1
Pid: 254144
Uptime: 0d 0h05m04s
Uptime_sec: 304
Memmax_MB: 0
PoolAlloc_MB: 281
PoolUsed_MB: 281
PoolFailed: 0
Ulimit-n: 3150919
Maxsock: 2250475
Maxconn: 900000
Hard_maxconn: 900000
CurrConns: 2827
CumConns: 69897
CumReq: 85073
MaxSslConns: 0
CurrSslConns: 0
CumSslConns: 15
Maxpipes: 225000
PipesUsed: 1957
PipesFree: 113
ConnRate: 170
ConnRateLimit: 0
MaxConnRate: 483
SessRate: 170
SessRateLimit: 0
MaxSessRate: 483
SslRate: 0
SslRateLimit: 0
MaxSslRate: 3
SslFrontendKeyRate: 0
SslFrontendMaxKeyRate: 3
SslFrontendSessionReuse_pct: 0
SslBackendKeyRate: 0
SslBackendMaxKeyRate: 0
SslCacheLookups: 3
SslCacheMisses: 3
CompressBpsIn: 0
CompressBpsOut: 0
CompressBpsRateLim: 0
ZlibMemUsage: 0
MaxZlibMemUsage: 0
Tasks: 8384
Run_queue: 1
Idle_pct: 99
node: haproxy-frontend
Stopping: 0
Jobs: 3090
Unstoppable Jobs: 1
Listeners: 260
ActivePeers: 0
ConnectedPeers: 0
DroppedLogs: 0
BusyPolling: 0
FailedResolutions: 0
TotalBytesOut: 270465011100
TotalSplicedBytesOut: 269255890440
BytesOutRate: 858092160
DebugCommandsIssued: 0
CumRecvLogs: 0
Build info: 3.0-dev12-d236b43
Memmax_bytes: 0
PoolAlloc_bytes: 294879040
PoolUsed_bytes: 294879040
Start_time_sec: 1716301289
Tainted: 0x3
TotalWarnings: 0
MaxconnReached: 0
BootTime_ms: 1980
Niced_tasks: 0
```

I could collect `show sess` and `show fd` like before if you want. And if there's any other info you need, please let us know.

**Comment by einval22** [comment]
Hi Felipe !

Thanks a lot for the feedback !

Could you please, test with the latest master and give us the results.

This patch on top  [5f713c0](https://github.com/haproxy/haproxy/commit/5f713c03bed6424fee7dc36fbb00efe368a7e95b) should resolve the problem.

Kind regards,

**Comment by felipewd** [comment]
hello @vkssv we've deployed the git tree up to [5f713c0](https://github.com/haproxy/haproxy/commit/5f713c03bed6424fee7dc36fbb00efe368a7e95b) and it does indeed fixed the problem, thanks!

Since this is a 3.0-dev10 regression, should we close this ticket?

**Comment by einval22** [comment]
Hi Felipe ! 

Yes, thanks a lot for testing ! You could close this, as it is not need to be backported.

**Comment by einval22** [comment]
Hi Felipe !

Here is an additional fix, which is very important: [83ab147](https://github.com/haproxy/haproxy/commit/83ab1479d0d30d50d55892d8733d1703a4ab22cc) 

Sorry for this, would you please take the latest master.

Kind regards,

**Comment by felipewd** [comment]
hi @vkssv we've deployed the latest master (including [83ab147](https://github.com/haproxy/haproxy/commit/83ab1479d0d30d50d55892d8733d1703a4ab22cc)) now.

Looking good, no leaks or issues.
